File: Data Structure/Hash/Easy/LET_no290_Word Pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        stringstream ss;
        ss << s;
        string cur;
        vector<string> words;

        while (ss >> cur)
            words.push_back(cur);

        if (words.size() != pattern.size())
            return false;

        unordered_map<char, string> ch_to_word;
        unordered_set<string> used_words;

        for (int i = 0; i < pattern.size(); i++) {
            char c = pattern[i];
            string cur = words[i];

            auto it1 = ch_to_word.find(c);
            auto it2 = used_words.find(cur);
            if (it1 != ch_to_word.end() && it1->second != cur || it1 == ch_to_word.end() && it2 != used_words.end())
                return false;

            ch_to_word[c] = cur;
            used_words.insert(cur);
        }

        return true;
    }
};

File: Data Structure/Hash/Easy/LET_no290_Word Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False
        
        memo, used = {}, set()
        for c, w in zip(pattern, words):
            if c not in memo:
                if w in used:
                    return False
                memo[c] = w
                used.add(w)
            elif memo[c] != w:
                return False
                
        return True